libraries/AP_OpticalFlow/AP_OpticalFlow_Pixart.cpp

```diff
@@ -166,25 +166,40 @@ bool AP_OpticalFlow_Pixart::setup_sensor(void)
 
     integral.last_frame_us = AP_HAL::micros();
 
-    _dev->register_periodic_callback(2000, FUNCTOR_BIND_MEMBER(&AP_OpticalFlow_Pixart::timer, void));
+    _dev->register_periodic_callback(5000, FUNCTOR_BIND_MEMBER(&AP_OpticalFlow_Pixart::timer, void));
     return true;
 }
 
 // write an 8 bit register
 void AP_OpticalFlow_Pixart::reg_write(uint8_t reg, uint8_t value)
 {
+#if CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BH
```

Comment:
Why is this difference needed? Our HAL API is supposed to be done in a way that drivers don't have to care about the underlying system; we should fix it if something isn't working as it should.

Reply:
This driver would be the first time used under Linux. In the beginning, I tried it wihout modification, not working, couldn't even read the PRODUCT_ID register.

Since I have a [CJMCU-3901](https://discuss.ardupilot.org/t/breakout-board-for-pmw3901-optical-flow-sensor/30613), so I tested it and found the existing driver code probably assumes that it's a 3-Wire SPI halfduplex device. However, mine seems to be a 4-Wire SPI fullduplex device according to its [datasheet](http://www.pixart.com/upload/POT0189-PMW3901MB-TXQT-DS-R1.10-200617%20(NonNDA)_20170720181537_20170802183402.pdf).

Then, I added the macro guard as you saw, to support fullduplexing, and, it works ('works' means seeing the expected OPTFLOW data from GCS). In advance, from the commit log of this driver, it was for PIXART 3900 at the very beginning, 3901 support was added several months after 3900. I can't tell what difference they have due to no info on 3900 at all.

But, compare the source code and 3901 register map, I think there is a difference though, the MOTION_BURST, 3900 and 3901 has different register map from `struct MotionBurst {}` and 3901 datasheet `Table 7. Register List`.

So it might not be the issue of HAL_Linux API, just different usage of 3900 and 3901.

This macro guard could be named e.g. `PIXART_PMW3901` or similar. I used `CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BH` because it was only tested on BH + RPI3. It still needs to be confirmed by other Boards and Systems. Titled WIP for the same reason.

Reply:
this makes no sense. We're using 4-wire SPI on all of the boards. Have you looked at what is happening with a logic analyser?

Reply:
No I have not seen it with an analyser, only tested using a RPI3 and CJMCU-3901. With the PR, it does work in my enviroment, but not without. Thanks for the reminding, I'll compare the two impl to see what happended.

@tridge BTW, do you have a PIXART 3900 datasheet since the code was originally committed by you? Seems no where could find it.  There is PMW3901MB-TXQT though, I doubt that 3900 and 3901 have the different register map, but the code treats them the same.